src/bvals/comms/coalesced_buffers.cpp

```diff
@@ -0,0 +1,482 @@
+
+  // Allocate and free buffers as required
+  int idx{0};
+  auto &stat = sparse_status_buffer.buffer();
+  for (auto uid : var_set) {
+    // Skip this variable if it is not communicated in this BoundaryType
+    if (coalesced_info_buf.count(uid) == 0) continue;
+    for (auto &[bnd_id, pvbbuf] : coalesced_info_buf.at(uid)) {
+      if (stat[idx] == 1) {
+        pvbbuf->SetReceived();
+        if (!pvbbuf->IsActive()) pvbbuf->Allocate();
+      } else {
+        pvbbuf->SetReceivedNull();
+        if (pvbbuf->IsActive()) pvbbuf->Free();
+      }
+      idx++;
+    }
+  }
+
+  auto &bids = GetBndIdsOnDevice(vars);
+  Kokkos::parallel_for(
+      PARTHENON_AUTO_LABEL,
+      Kokkos::TeamPolicy<>(parthenon::DevExecSpace(), bids.size(), Kokkos::AUTO),
+      KOKKOS_LAMBDA(parthenon::team_mbr_t team_member) {
+        const int b = team_member.league_rank();
+        if (bids[b].buf_allocated) {
+          const int buf_size = bids[b].size();
+          Real *com_buf = &(bids[b].coalesced_buf(bids[b].start_idx()));
+          Real *buf = &(bids[b].buf(0));
+          Kokkos::parallel_for(Kokkos::TeamThreadRange<>(team_member, buf_size),
+                               [&](const int idx) { buf[idx] = com_buf[idx]; });
+        }
+      });
+  coalesced_comm_buffer.Stale();
+  sparse_status_buffer.Stale();
+
+  return true;
+}
+
+//----------------------------------------------------------------------------------------
+void CoalescedBuffer::AddVarBoundary(BndId &bnd_id) {
+  auto key = GetChannelKey(bnd_id);
+  PARTHENON_REQUIRE(pmesh->boundary_comm_map.count(key), "Buffer doesn't exist.");
+  var_buf_t *pbuf = &(pmesh->boundary_comm_map.at(key));
+  coalesced_info_buf[bnd_id.var_id()].push_back(std::make_pair(bnd_id, pbuf));
+  current_size += bnd_id.size(); // This will be the maximum size of communication since
+                                 // it includes all variables
+  all_vars.insert(bnd_id.var_id());
+}
+
+void CoalescedBuffer::AddVarBoundary(MeshBlock *pmb, const NeighborBlock &nb,
+                                     const std::shared_ptr<Variable<Real>> &var) {
+  // Store both the variable-boundary buffer information and a pointer to the v-b buffer
+  // itself associated with var ids
+  BndId bnd_id = BndId::GetSend(pmb, nb, var, b_type, partition, -1);
+  AddVarBoundary(bnd_id);
+}
+
+//----------------------------------------------------------------------------------------
+//----------------------------------------------------------------------------------------
+//----------------------------------------------------------------------------------------
+CoalescedBuffersRank::CoalescedBuffersRank(int o_rank, BoundaryType b_type, bool send,
+                                           mpi_comm_t comm, Mesh *pmesh)
+    : other_rank(o_rank), b_type(b_type), sender(send), buffers_built(false), comm_(comm),
+      pmesh(pmesh) {
+
+  int tag = 1234 + static_cast<int>(GetAssociatedSender(b_type));
```

Comment:
What's the `1234` doing here?

Reply:
just an arbitrary tag.

Reply:
why is an arbitrary tag needed as an offset for the `GetAssociatedSender`?

Reply:
I second the question (suggestion for comment in code) of this magic number.

Reply:
I was initially sending on `COMM_WORLD` I think so I just chose something other than zero to minimize the likelihood of a collision. Probably doesn't need to be offset anymore, so just removed.